packages/types/index.d.ts

```diff
@@ -354,6 +354,7 @@ export interface IActionRunResult {
   nextStep?:
     | { command: 'jump-to-step'; stepId: IStep['id'] }
     | { command: 'stop-execution' }
+  nextStepMetadata?: IJSONObject
```

Comment:
Very nervous about typing this as `IJSONObject`, which is no better than `any`. Is there a way we can improve type safety on this

Reply:
We also need to consider for-each, which will require 

I'm thinking tagged union or similar, e.g.

```
interface PostmanSendEmailMetadata {
  // have metadataType as distinguisher..?
  metadataType: 'postman-send-email'
  progress: number
}

interface TilesQueryRowsMetadata {
  metadataType: 'tiles-query-rows'
  numRowsRemaining: number
}

type NextStepMetadata = PostmanSendEmailMetadata | TilesQueryRowsMetadata //.. etc

interface Metadata {
  // Common / core metadata fields
  core: {
    forEachIndex: number | null
  }
  
  // Next-step-specific metadata.
  stepSpecific: NextStepMetadata
}
```

Might be too elaborate, maybe there's a simpler approach.

Reply:
Let's go with union

Reply:
union seems good, but the down side is that we're handling the case of missing metadata anyway (as the whole object is optional). Is it still really beneficial to use union type and have to deal with incorrect metadata object (e.g. query tile metadata used on postman) specifically, or rather just collapse it into the missing data handling?

Reply:
hmmm... if I'm understanding your concern right, typescript should block accessing anything in the `stepSpecific` object until we check `stepSpecific.metadataType`. So there won't be an issue with app A querying the wrong metadata typ

I'm generally more preferential to stronger types because i know i screw up a lot without strong typing. old man cannot remember what the types are in my head anymore.

Reply:
@ogp-weeloong yup that's right. And the downside is that now we have to manage multiple types for different kinda actions (Might end up doing weird casting so I don't think it's much better than just `IJSONObject` and be disciplined about handling missing data

Reply:
Paiseh, going to try pushing back a little bit more (as I'm personally a strong proponent of static typing) :P Do you have an example where we will end up in a scenario where we will do weird casting?

My main concern with IJSONObject is that:
*  Discipline doesn't scale to large teams, and it keeps product tech knowledge in the team's oral history, which is bad for knowledge management (unless we're heavily commenting code. But if we're commenting, we might as well convert our comments to static types).
* We're forcing ourselves to always perform casting if we stick with IJSONObject, whereas with stronger types, we at least have some scenarios where we don't have to cast and can leverage the type checker.
* I am also worried loose typing will very easily cause explosions if we ever do metadata changes (e.g. we want to change the metadata struct , but we still have lots of jobs with the old metadata in the queue. It's very very easy to forget what's metadata is in the old jobs unless we document it down)

My above approach is probably over complex, but I feel like something simpler will also be ok. e.g. even just
```
interface Metadata {
  progress: number
}
```

Reply:
Had a discussion.
Adding typings